### PR TITLE
Game Loop Pattern

### PR DESCRIPTION
**Pull request by Azureyjt**
This PR should fixed issue #551 

Implement Game Loop pattern.

### CODE REVIEW

**Comment by Azureyjt** [comment]
@iluwatar Please kindly review this PR.

File: game-loop/src/test/java/com/iluwatar/gameloop/BulletTest.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    try {
+      bullet.setPosition(1.5f);
+      var field = Bullet.class.getDeclaredField("position");
+      field.setAccessible(true);
+      Assert.assertEquals(1.5f, (Float) field.get(bullet), 0);
+    } catch (NoSuchFieldException | IllegalAccessException e) {
+      Assert.fail("Fail to modify field access.");
+    }
+  }
+}
```
Comment (iluwatar) [inline game-loop/src/test/java/com/iluwatar/gameloop/BulletTest.java]: I don't think these Bullet tests add very much value. It's a simple class with one setter and getter and as such doesn't need to be separately tested. Another thing is the approach to testing where private fields are made accessible. It's not good practise and we should try to find alternate methods.

File: game-loop/src/test/java/com/iluwatar/gameloop/VariableStepGameLoopTest.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  public void tearDown() {
+    gameLoop = null;
+  }
+
+  @Test
+  public void testUpdate() {
+    try {
+      var method = VariableStepGameLoop.class.getDeclaredMethod("update", Long.class);
+      method.setAccessible(true);
+      method.invoke(gameLoop, 20L);
```
Comment (iluwatar) [inline game-loop/src/test/java/com/iluwatar/gameloop/VariableStepGameLoopTest.java]: I don't think it's good practise use reflection to make methods accessible. Maybe we could just make the method public by default? Or if that's a problem from design perspective maybe inherit from VariableStepGameLoop in this test and override the method visibility. The same issue is present in other tests too.

**Comment by Azureyjt** [comment]
Thank you for your suggestion iluwatar. I'm on it.

File: game-loop/src/main/java/com/iluwatar/gameloop/App.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+  /**
+   * Program entry point.
+   * @param args runtime arguments
+   */
+  public static void main(String[] args) {
+
+    try {
+      LOGGER.info("Start frame-based game loop:");
+      GameLoop gameLoop = new FrameBasedGameLoop();
```
[the remark this answers is not in the document]
Comment (Azureyjt) [reply game-loop/src/main/java/com/iluwatar/gameloop/App.java]: Fixed in commit a0f8a6152eabb1909157baf32319cbe597e1d13f

File: game-loop/src/main/java/com/iluwatar/gameloop/GameController.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+package com.iluwatar.gameloop;
+
+/**
+ * Update and render objects in the game. Here we add a Bullet object to the
+ * game system to show how the game loop works.
+ */
+public class GameController {
+
+  private Bullet bullet;
```
[the remark this answers is not in the document]
Comment (Azureyjt) [reply game-loop/src/main/java/com/iluwatar/gameloop/GameController.java]: Fixed in commit a0f8a6152eabb1909157baf32319cbe597e1d13f

File: game-loop/src/main/java/com/iluwatar/gameloop/GameLoop.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+import org.slf4j.LoggerFactory;
+
+/**
+ * Abstract class for GameLoop implementation class.
+ */
+public abstract class GameLoop {
+
+  protected final Logger logger = LoggerFactory.getLogger(this.getClass());
+
+  protected GameStatus status;
```
[the remark this answers is not in the document]
Comment (Azureyjt) [reply game-loop/src/main/java/com/iluwatar/gameloop/GameLoop.java]: Fixed in commit a0f8a6152eabb1909157baf32319cbe597e1d13f

File: game-loop/src/test/java/com/iluwatar/gameloop/BulletTest.java (hunk 1 above)
Comment (Azureyjt) [reply game-loop/src/test/java/com/iluwatar/gameloop/BulletTest.java]: Fixed in commit de763fbc4f138c37212e81e61710c8828e5f43cc

File: game-loop/src/test/java/com/iluwatar/gameloop/VariableStepGameLoopTest.java (hunk 2 above)
Comment (Azureyjt) [reply game-loop/src/test/java/com/iluwatar/gameloop/VariableStepGameLoopTest.java]: Fixed in commit de763fbc4f138c37212e81e61710c8828e5f43cc

File: game-loop/README.md [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+tags:  
+ - Java  
+ - Difficulty-Beginner  
+---  
+  
+## Intent  
+A game loop runs continuously during gameplay. Each turn of the loop, it processes user input without blocking, updates the game state, and renders the game. It tracks the passage of time to control the rate of gameplay.
+
+This pattern decouple the progression of game time from user input and processor speed.
+
```
[the remark this answers is not in the document]
Comment (Azureyjt) [reply game-loop/README.md]: Fixed in commit ffcdca53ef4e4c0a7935ffd913ebdf508a5fbef5

**Comment by iluwatar** [comment]
This looks good @Azureyjt. Thanks once again for a great contribution!

**Pull request merged**